### PR TITLE
Add calypso-razorpay package

### PR DESCRIPTION
**Pull request by nbloomf**
We're preparing to add support for the Razorpay payment processor to calypso checkout. This PR introduces a package, calypso-razorpay, that wraps the Razorpay JS library in a react hook, analogous to the existing [calypso-stripe](https://github.com/Automattic/wp-calypso/tree/trunk/packages/calypso-stripe) package.

See p2-pbOQVh-4mH for more details.

## Proposed Changes

* Adds a package, analogous to calypso-stripe, for interacting with Razorpay from inside calypso.

## Testing Instructions

* See #86197

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by nbloomf** [comment]
If the "Needs Review" tag doesn't actually mean "I want this reviewed", but rather "I want canary tests to run on this PR", maybe we should rename it.

File: packages/calypso-razorpay/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	options: { key: string };
+}
+
+export type Razorpay = object;
+
+export type RazorpayLoadingError = undefined | null | Error;
+
+export interface RazorpayData {
+	razorpay: Razorpay | null;
+	razorpayConfiguration: null | RazorpayConfiguration;
```
Comment (ignaciodob) [inline packages/calypso-razorpay/src/index.tsx]: Nitpicking, but it bothers me that we are not consistent in our placement of `null` here and in the line above 
```
	razorpay: Razorpay | null;
	razorpayConfiguration: null | RazorpayConfiguration;
```

File: packages/calypso-razorpay/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+const RazorpayContext = createContext< RazorpayData | undefined >( undefined );
+
+export interface UseRazorpayJs {
+	razorpay: Razorpay | null;
+	isRazorpayLoading: boolean;
+	razorpayLoadingError: RazorpayLoadingError;
+}
+
+export type GetRazorpayConfigurationArgs = { sandbox: boolean }; // TODO: fix this dummy argument
```
Comment (ignaciodob) [inline packages/calypso-razorpay/src/index.tsx]: There's a TODO here, but I think it's addressed already

File: packages/calypso-razorpay/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+				setRazorpayConfigurationError( error );
+			} );
+		return () => {
+			isSubscribed = false;
+		};
+	}, [ memoizedRequestArgs, fetchRazorpayConfiguration ] );
+	return { razorpayConfiguration, razorpayConfigurationError };
+}
+
+// TODO
```
Comment (ignaciodob) [inline packages/calypso-razorpay/src/index.tsx]: Bringing attention to this TODO

File: packages/calypso-razorpay/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+function areRequestArgsEqual(
+	previous: undefined | null | GetRazorpayConfigurationArgs,
+	next: undefined | null | GetRazorpayConfigurationArgs
+): boolean {
+	return !! previous && !! next; // TODO
```
Comment (ignaciodob) [inline packages/calypso-razorpay/src/index.tsx]: Bringing attention to this TODO

File: packages/calypso-razorpay/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+ */
+export function useRazorpay(): RazorpayData {
+	const razorpayData = useContext( RazorpayContext );
+	if ( ! razorpayData ) {
+		throw new Error( 'useRazorpay can only be used inside a RazorpayHookProvider' );
+	}
+	return razorpayData;
+}
+
+// See https://usehooks.com/useMemoCompare/
```
Comment (ignaciodob) [inline packages/calypso-razorpay/src/index.tsx]: That URL doesn't seem to work anymore and I don't see see it in the usehooks directory either 🤔

**Review by ignaciodob — COMMENTED** [review]
Overall, this is looking good to me, but we should probably get some reviews from the lovely folks at @Automattic/shilling too

**Review by sirbrillig — COMMENTED** [review]
I pushed a small change which declares a Razorpay type on the window object so we don't need to have those `ts-ignore` comments.

File: packages/calypso-razorpay/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+import { useRef, useEffect, useState, useContext, createContext, PropsWithChildren } from 'react';
+
+const debug = debugFactory( 'calypso-razorpay' );
+
+export interface RazorpayConfiguration {
+	js_url: string;
+	options: { key: string };
+}
+
+declare class Razorpay {
```
Comment (sirbrillig) [inline packages/calypso-razorpay/src/index.tsx:11]: Stripe.js used to be hot-loaded this way too before https://github.com/Automattic/wp-calypso/pull/56089. 

If you need later to explain what the `Razorpay` object can do, you can add things to this type. Here's what we used to do for Stripe: https://github.com/Automattic/wp-calypso/blob/c38fd60dadca86a0bb09d684682f4a964ab63ddc/packages/calypso-stripe/src/index.tsx#L44-L58

File: packages/calypso-razorpay/src/index.tsx (hunk 4 above)
Comment (sirbrillig) [reply packages/calypso-razorpay/src/index.tsx]: You might want to actually compare the properties of these arguments to make sure that the memoization will change.

File: packages/calypso-razorpay/src/index.tsx (hunk 5 above)
Comment (sirbrillig) [reply packages/calypso-razorpay/src/index.tsx]: Interesting. I guess they must have removed it? Well, we've been using this pattern successfully for a while. We might just want to remove the comment.

**Review by sirbrillig — COMMENTED** [review]
Overall this looks good to me! I think the configuration comparison function needs to be updated to actually compare two options but otherwise I have no comments.

**Review by sirbrillig — APPROVED** [review]

**Comment by nbloomf** [comment]
The failing e2e tests are unrelated; these are affecting other PRs as well.

**Pull request merged**